### ISSUE TITLE
Add search bar at the top

### ISSUE THREAD

**Issue by samyue888** [ocupop/librestream#90]
![image](https://user-images.githubusercontent.com/81642726/115453780-d8f4e980-a1d4-11eb-967a-47135edd30a9.png)

Same function as old site:
https://librestream.kinsta.cloud/

But please exclude all /form-confirmations/ and support pages

**Comment by tbeck** [comment]
@samyue888 - Do we have finalized design direction for us to work from on this? I saw the rough designs that Chuck provided in email. Could you post designs here if they are approved and we can begin working to implement?

Thanks!

**Comment by samyue888** [comment]
@tbeck here is the final design:
https://www.dropbox.com/s/83r8yjltz24m80e/2827-LIB_website-2021_phase-2-Search_d2.pdf?dl=0

Here are the specs:

Note that the search bar, when expanded, should cover Login and Contact.

Search should include these pages (home, platform pages & subpages, use cases, solutions, about) and posts only.

Search within title and body text.

If search phrase is in title, highlight the phrase. 

If search phrase is in body text, highlight the phrase and show the first instance in the result description. If phrase is not in body text, show the first sentence in the result description. 

List results in descending order of number of matches.

![image](https://user-images.githubusercontent.com/81642726/116744784-9e970380-a9af-11eb-80db-397ddd67f780.png)

**Comment by paulstroot** [comment]
Hi @samyue888 , I've been doing some work with the search feature on Librestream and I've hit a point where we need some feedback from your team on how to proceed.

With the method we have been using for generating search results at this point, all of the search data is indexed into a data file to to quickly return results. This was working great on the news pages where just posts were being returned. When I added pages, case studies and solutions, we hit a limit to the size of the data that can be returned. We get an error "[Response Object too large](https://librestreamdevelopsearch.gatsbyjs.io/static/eaa5526911a48b9208c4e6f1c8ce242f.index.txt)".

There are a couple of directions to go here. 

1. We could split up the search into different types, and return "post results", "page results", "Case studies", etc. This should get around our error message, and be the fastest solution, but we would not be able to mix the search results into one list sorted by relevance. Plus, there is no guarantee that we might not hit our limit in the future as we add more content to the site.

2.  The other solution would be to load the search results in at runtime from a server. One way to do this would be to query our Wordpress CMS directly. The other way would be to use a third party search service such as "Algolia". I'll explain some pros and cons of these two here so you can help make a choice.

 - Using a third party search service such as [Algolia](https://www.algolia.com/) gives a great amount of control over the search. Their dashboard allows you to give weight to different types of content to customize the order in which the search results appear. It's quick, and can pretty easily update the results as you type in the search field. There is a [monthly cost](https://www.algolia.com/pricing/). (Note: Algolia is the service which I have used before and am familiar with. There are other third-party search solutions out there if we need to research others)
 
- Qurying our Wordpress CMS would be a little easier, since it does not introduce any new services. We would have less control over the fine-tuning of how search results appear, and it would probably be a little bit slower, cause more of a delay in the "results as you type".

This unfortunate discovery of our size limits on our search will also result in more time to get this working properly. I think we will need this week to explore and plan out an appropriate solution, next week to implement, and the following week to fine-tune and test.

cc: @tbeck

**Comment by paulstroot** [comment]
@samyue888 - I've discussed a little bit with Tom and we think that using the third party tool "Algolia" is the best approach here. I've done a proof of concept so you can see how quickly it works, and the nice way it highlights the search terms and pulls relevant snippets out of the content.
[https://librestreamdevelopsearch.gatsbyjs.io/algoliasearch/](https://librestreamdevelopsearch.gatsbyjs.io/algoliasearch/)
(note that only posts and pages are being displayed, and we will want to work on the search algorithm a little bit)

Some great features about Algolia is that you can really dial in the search results by doing thigns like giving different priority to different content types, setting typo tolerance, setting up synonyms and word proximity and other features. 

**Before going any further though we will need your approval.** There is an associated cost with this service, but I believe it is not much. (@tbeck may have details on this, as he's worked with it on other sites). I would also like to discuss the best way to mix search results for different types of content. (pages, posts, solutions, etc)

**Comment by samyue888** [comment]
@paulstroot The search bar in that link is not working (ie. when I type anything in, it shows 0 results), so I cannot test and verify if the design is correct. We would need to review a working version before proceeding.

Please make sure the following points are addressed, as per my message above:
- Search should include these pages (home, platform pages & subpages, use cases, solutions, about) and posts only.
- Search within title and body text.
- If search phrase is in title, highlight the phrase.
- If search phrase is in body text, highlight the phrase and show the first instance in the result description. If phrase is not in body text, show the first sentence in the result description.

Also, from the link you sent, I noticed a lot of strange text in the results:
![image](https://user-images.githubusercontent.com/81642726/119873816-c8bde180-bed9-11eb-82e2-ff1a3cd40608.png)
![image](https://user-images.githubusercontent.com/81642726/119873849-d3787680-bed9-11eb-8d40-12b55edea34d.png)

It is also pulling confirmation pages, which we do not want to include:
![image](https://user-images.githubusercontent.com/81642726/119874052-11759a80-beda-11eb-971d-4505af9fc7f6.png)

**Comment by paulstroot** [comment]
Thanks @samyue888 . Don't worry about the details too much now. This is just a proof-of-concept to demonstrate the Algolia solution. Strange that the search was not working for you though. I tried it on several devices and it is updating the search as I type. Have you tried a different browser, device or incognito window?

**Comment by samyue888** [comment]
@paulstroot Our previous developer informed us that the old site used this plugin https://searchwp.com

**Comment by paulstroot** [comment]
I can check it out. Do you know if you have a license purchased for it
<https://searchwp.com/buy/>?

**Comment by samyue888** [comment]
Still waiting for their reply. In the meantime, I think we can try out the built-in WP search that you had mentioned.

**Comment by paulstroot** [comment]
I have the search working on the [gatsby-side](https://librestreamdevelopsearch.gatsbyjs.io/search) of the site, pulling from the WordPress REST API. (With some smaller tweaks and design updates still needed). 

The next task would be to modify the WordPress CMS side of it to make sure that it properly searches all of the correct content from the different page types, excludes the ones we do not want, and highlights the proper excerpts. I'm going to wait on this until you can find out if we want to use [SearchWP ](https://searchwp.com/buy/) @samyue888 . It looks like that might be the perfect tool for this, and I don't want to spend a lot of development time now if we might be re-doing it with SearchWP if we decide to go down that road. Let me know when you find out about the license.

FYI, I can build this without SearchWP, but we will not have as much control over the search, and it make take more development time.

**Comment by paulstroot** [comment]
Any word on this yet @samyue888 ?

**Comment by samyue888** [comment]
Not yet. I've sent a followup this morning asking if we have a subscription. In the meantime, could you proceed with the free WP suggestion? Thanks!

**Comment by samyue888** [comment]
@paulstroot Just got a reply. Our licence has expired

**Comment by paulstroot** [comment]
Does that mean that you would not like to use that tool? Or would you like
to get a new license?

**Comment by samyue888** [comment]
We are leaning towards not using SearchWP, but could you please list out the limitations of the free WP option?

**Comment by paulstroot** [comment]
If we do not use a third-party plug-in like Search WP, I will need to write
the search code myself. The search will likely be slower, you will not be
able to configure anything through the wordpress admin, we may have limited
options regarding how the excerpt is pulled out of the content with the
highlighted word, there will not be any ordering of results based on
matching. There may be a couple other things we come across as we begin
working with it.

**Comment by samyue888** [comment]
Thanks. Let me discuss with @lsdrew and get back to you

**Comment by samyue888** [comment]
After some internal discussion, lets go with Algolia - it seems to be the cheapest and what you're most familiar with.

CC @lsdrew

**Comment by paulstroot** [comment]
@samyue888 - Algolia it is. I have a great start on it already. I'm working with an Ocupop development account for now, but we'll need you to set up a new Algolia account and either share login credentials, or add dev@ocupop.com as a user.

**Comment by paulstroot** [comment]
Hi @samyue888 , Ready for first round of review on the search page. Here is the link:
https://librestreamdevelopsearch.gatsbyjs.io/search/

**Comment by samyue888** [comment]
@paulstroot Looks great so far - thanks! I'm testing with 'remote' as a search term and for the 1st result description, we somehow have the text duplicated 3 times. Would you mind troubleshooting:
![image](https://user-images.githubusercontent.com/81642726/125687661-117cf2f1-fde1-4a04-b573-167134c045ad.png)
https://librestreamdevelopsearch.gatsbyjs.io/search?s=remote

**Comment by samyue888** [comment]
Also, there's an overlap of the magnifying glass and X here:
![image](https://user-images.githubusercontent.com/81642726/125687826-8b0decfa-ac10-45df-a2f3-8b369ff1f6c7.png)

And when I click load more results, it takes me to a blank page:
![image](https://user-images.githubusercontent.com/81642726/125687944-c62ec57f-a229-46be-9659-2cb10246be99.png)

**Comment by paulstroot** [comment]
I can take a look at that. What's happening is that the same text is
probably entered in for the content, excerpt and the meta description, and
the search is showing all three results because there is a match in each of
the three string. This is probably an update from within Algolia and I'll
see if I can update that tomorrow.

**Comment by samyue888** [comment]
@paulstroot As per our call, please ensure the enter button and magnifying glass works for the search bar on the page, and change the yellow highlight to a blue font.

I have just created an Algolia account. What would be the next step?

**Comment by paulstroot** [comment]
Looking at the design here: https://www.dropbox.com/s/83r8yjltz24m80e/2827-LIB_website-2021_phase-2-Search_d2.pdf?dl=0, it looks like the highlighted words should be light blue and underlined. I that correct?

Can you share your access to your algolia account with dev@ocupop.com?

**Comment by samyue888** [comment]
@paulstroot The comp has it underlined, but can we try without the underline? I think it would look better. 

I've sent an invite from Algolia.

**Comment by paulstroot** [comment]
Thanks @samyue888 - Is there a higher level of permissions you are able to grant us for Algolia? There are some API keys that are only available to view to users with the proper permissions.

**Comment by samyue888** [comment]
@paulstroot Please check now. Thanks

**Comment by paulstroot** [comment]
@samyue888 - Got it! I've updated the page, and it is now pulling data from your Algolia account. Please review.
https://librestreamdevelopsearch.gatsbyjs.io/search/?s=remote

I encourage you to browse through your Algolia account and look at all of the configuration settings for refining the search. There are a lot of great options to fine-tune the results that show up to your liking. Start by going to "indicies", then select the index "LibrestreamSite", and you can start taking a look at the settings in that section.

If you notice any pages showing up in the results that should not be, you may need to go into that page in the CMS and make sure "Allow search engines to show this Page in search results?" is set to "NO". This is at the bottom of the edit page under "Yoast SEO -> Advanced". 

Changes to the Algolia Search get re-indexed every time the site is rebuilt. On the staging site this will be every time a page is saved. On development you may need to re-build in gatsby cloud for your WordPress pages to be re-indexed in Algolia.

**Comment by samyue888** [comment]
Thanks @paulstroot After reviewing the results, what are the next steps for us?

**Comment by samyue888** [comment]
Also, the teal color is too bright. Could we change it to the blue as specified in the comp?

**Comment by paulstroot** [comment]
Thanks @samyue888 - these two items should be fixed now.

**Comment by samyue888** [comment]
@paulstroot We've reviewed the settings and added our payment method. Please add a PR when ready, thanks!

**Comment by paulstroot** [comment]
Sure thing. I'll do this on Monday just to make sure I'm available in case
anything needs updating.

**Comment by paulstroot** [comment]
@samyue888 - I have merged this branch into develop and I'm about to open a pull request. FYI, there are a couple tests that do not pass right now in our Automated tests. The first one is with our contact form, which we have already discussed. The second one is a broken link on the Onsight Support Page. This is the User Manual Link under "Onsight Hub v2019.3", and the error is because of the redirect issue that we are experiencing, and that Gatsby Cloud is supposedly fixing soon.

**Comment by samyue888** [comment]
Thanks. @paulstroot regarding the user manual not loading, I noticed most of the PDFs in the support/training pages are static links. Why is this so? It would be nice to have this "Onsight Hub v2019.3" as static too, to avoid the redirect issue.

**Comment by paulstroot** [comment]
Good observation @samyue888 . In the CMS admin, this link was set as "external link" which goes to a page which then redirects to the PDF. I'm not sure what the decision was to point to that link instead of the PDF. I updated this to point directly to the PDF and it is working on [the staging site](https://librestreamstaging.gatsbyjs.io/onsight-support/#cG9zdDo3NTU=) now.

When you publish the production site in Gatsby cloud it should fix that link there also. _(Note: I needed to click "Clear Cache and Bulid Main" in order to make it work properly, see the attached image.)_

![image](https://user-images.githubusercontent.com/52838369/128027633-cc35b352-dca7-443e-800d-958ff1a1aa08.png)



![image](https://user-images.githubusercontent.com/52838369/128022625-2b10984b-446e-4e9e-8819-429dcce30be3.png)
